text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+
+Glimmer components will only receive the owner directly, and as such will _not_
+support type injections. This cuts down on the implicit knowledge developers
+must have when writing a component.
+
+### Properties
+
+Glimmer components have 3 properties: `args`, `isDestroying`, and `isDestroyed`.
+
+#### `args`
+
+As discussed in the motivation section, `args` is an object with the values of
+the named arguments passed to the component. This property will be updated
+whenever the arguments change, with the `didUpdateArgs` hook being called after
+any updates. It will be shallowly frozen in development mode to prevent users
+from setting values on it.
+
+#### `isDestroying`
+
+This property will be set to `true` when component teardown has been initiated,
+_before_ the component's `willDestroy` hook is run, along with any other
+components which are currently being torn down. This allows the entire component
+tree to be marked before user code is run. It can be used by users to
+conditionally prevent asynchronous code from running, and to check on the
+teardown state of the component in general.
+
+#### `isDestroyed`
+
+This property will be set after any `willDestroy` hooks have run, and the
+component has been fully torn down. It can be used by users to conditionally
+prevent asynchronous code from running, and to check on the teardown state of
+the component in general.
+
+### Lifecycle Hooks
+
+Classic components have 13 major lifecycle hooks that run during 3 major phases
+of the component lifecycle, with some hooks running during multiple phases:
+
+1. **Init and Initial Render:**
+   * `init`
+   * `willInsertElement`
+   * `didInsertElement`,
+   * `didReceiveAttrs`
+   * `willRender`
+   * `didRender`.
+
+2. **Rerenders and Updates:**
+   * `didReceiveAttrs`
+   * `didUpdateAttrs`
+   * `willUpdate`
+   * `didUpdate`
+   * `willRender`
+   * `didRender`
+
+3. **Destruction:**
+   * `willDestroyElement`
+   * `didDestroyElement`
+   * `destroy`
+   * `willDestroy`
+
+Many of these hooks have overlapping or redundant functionality, and it's fairly
+confusing when to use which and what the differences are. We can simplify this
+cycle in a number of ways:
+
+* Hooks that run during multiple phases such as `didRender` and
+  `didRecieveAttrs` can be convenient at times, but also add mental overhead and
+  redundancy. We can remove these in favor of clearly delineated hooks which
+  only run during _one_ phase.
+
+* "Bookend" methods (`did*` and `will*`) can be confusing, since they require
+  some specific knowledge of what the "bookended" functionality is. For
+  instance, users almost _always_ want to use `didInsertElement` and
+  `willDestroyElement`, but the existence of their opposite bookends can make
+  this confusing. Additionally, the fact that `didReceiveAttrs` and
+  `didUpdateAttrs` do _not_ have opposing bookends is inconsistent with this
+  pattern.
```

Comment:
For me, the biggest motivation is to nudge people away from a programming model where lots of work is done eagerly, imperatively in response to events. This is really about moving components to a pull-based model where computation happens as a side-effect of its use, in addition to avoiding the runtime overhead of having to check for and invoke so many hooks. It adds up.

Reply:
This is true, but I think people may shy away from this line of reasoning because it's "what we had before" the hooks, and it was painful at times. I think pointing out the amount of redundancy is a stronger argument because it means nobody is "losing" anything, but I'll add another bullet point which discusses the benefits of using getters over imperative code in hooks.